notebooks/227-whisper-subtitles-generation/227-whisper-subtitles-generation.ipynb

```diff
@@ -0,0 +1,900 @@
+    "\n",
+    "\n",
+    "def decoder_forward(decoder, x: torch.Tensor, xa: torch.Tensor, kv_cache: Optional[dict] = None):\n",
+    "    \"\"\"\n",
+    "    Override for decoder forward method.\n",
+    "    Parameters:\n",
+    "      x: torch.LongTensor, shape = (batch_size, <= n_ctx) the text tokens\n",
+    "      xa: torch.Tensor, shape = (batch_size, n_mels, n_audio_ctx)\n",
+    "           the encoded audio features to be attended on\n",
+    "      kv_cache: Dict[str, torch.Tensor], attention modules hidden states cache from previous steps \n",
+    "    \"\"\"\n",
+    "    offset = next(iter(kv_cache.values())).shape[1] if kv_cache else 0\n",
+    "    x = decoder.token_embedding(\n",
+    "        x) + decoder.positional_embedding[offset: offset + x.shape[-1]]\n",
+    "    x = x.to(xa.dtype)\n",
+    "\n",
+    "    for block in decoder.blocks:\n",
+    "        x, kv_cache = block(x, xa, mask=decoder.mask, kv_cache=kv_cache)\n",
+    "\n",
+    "    x = decoder.ln(x)\n",
+    "    logits = (\n",
+    "        x @ torch.transpose(decoder.token_embedding.weight.to(x.dtype), 1, 0)).float()\n",
+    "\n",
+    "    return logits, kv_cache\n",
+    "\n",
+    "\n",
+    "# override decoder forward\n",
+    "model.decoder.forward = partial(decoder_forward, model.decoder)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "tokens = torch.ones((5, 3), dtype=torch.int64)\n",
+    "\n",
+    "logits, kv_cache = model.decoder(tokens, audio_features, kv_cache={})\n",
+    "kv_cache = {k: v for k, v in kv_cache.items()}\n",
+    "tokens = torch.ones((5, 1), dtype=torch.int64)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "outputs = [f\"out_{k}\" for k in kv_cache.keys()]\n",
+    "inputs = [f\"in_{k}\" for k in kv_cache.keys()]\n",
+    "dynamic_axes = {\n",
+    "    \"tokens\": {0: \"beam_size\", 1: \"seq_len\"},\n",
+    "    \"audio_features\": {0: \"beam_size\"},\n",
+    "    \"logits\": {0: \"beam_size\", 1: \"seq_len\"}}\n",
+    "dynamic_outs = {o: {0: \"beam_size\", 1: \"prev_seq_len\"} for o in outputs}\n",
+    "dynamic_inp = {i: {0: \"beam_size\", 1: \"prev_seq_len\"} for i in inputs}\n",
+    "dynamic_axes.update(dynamic_outs)\n",
+    "dynamic_axes.update(dynamic_inp)\n",
+    "torch.onnx.export(\n",
+    "    model.decoder, {'x': tokens, 'xa': audio_features, 'kv_cache': kv_cache},\n",
+    "    'whisper_decoder.onnx',\n",
+    "    input_names=[\"tokens\", \"audio_features\"] + inputs,\n",
+    "    output_names=[\"logits\"] + outputs,\n",
+    "    dynamic_axes=dynamic_axes\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Decoder model autoregressivly predict next token gaided by encoder headen states and previosly predixted sequence. \n",
```

Comment:
```suggestion
    "The decoder model autoregressively predicts the next token guided by encoder headen states and previously predicted sequence. \n",
```
Seems to be a few misspellings here. @eaidova can you say what "encoder headen" is supposed to mean? Please double check my corrections as well.